main_test.go

```diff
@@ -0,0 +1,267 @@
+	commandSpan      = `//span[contains(@class, "command")]`
+)
+
+var (
+	pw       *playwright.Playwright
+	browsers []playwright.BrowserType
+)
+
+func openChat(t *testing.T, browser playwright.Browser, username string) (playwright.Page, error) {
+	page, err := browser.NewPage()
+	if err != nil {
+		return nil, fmt.Errorf("could not open new browser page: %w", err)
+	}
+
+	_, err = page.Goto(url)
+	if err != nil {
+		return nil, fmt.Errorf("could not open chat page: %w", err)
+	}
+
+	_, err = page.WaitForSelector(nameInput)
+	if err != nil {
+		return nil, fmt.Errorf("could not find name input: %w", err)
+	}
+
+	err = page.Type(nameInput, username)
+	if err != nil {
+		return nil, fmt.Errorf("an error occured when typing the username: %w", err)
+	}
+
+	err = page.Click(joinButton)
+	if err != nil {
+		return nil, fmt.Errorf("could not click join button: %w", err)
+	}
+
+	_, err = page.WaitForSelector(msgOuputDiv)
+	if err != nil {
+		return nil, fmt.Errorf("chat window did not show: %w", err)
+	}
+
+	_, err = page.WaitForSelector(fmt.Sprintf(nameSpan, username))
+	if err != nil {
+		return nil, fmt.Errorf("could not get join message span: %w", err)
+	}
+
+	return page, nil
+}
+
+func sendMessage(page playwright.Page, msg string) error {
+	err := page.Type(msgInputTextarea, msg)
+	if err != nil {
+		return fmt.Errorf("could not find msg box: %w", err)
+	}
+
+	err = page.Click(sendInput)
+	if err != nil {
+		return fmt.Errorf("could not send message: %w", err)
+	}
+
+	return nil
+}
+
+func runTests(t *testing.T, f func(t *testing.T, browser playwright.Browser)) {
+	for _, bt := range browsers {
+		browser, err := bt.Launch()
+		if err != nil {
+			t.Errorf("could not launch browser: %v", err)
+		}
+		defer func() {
+			if browser.IsConnected() {
+				browser.Close()
+			}
+		}()
+
+		t.Run(bt.Name(), func(t *testing.T) {
+			f(t, browser)
+			browser.Close()
+		})
+	}
+}
+
+func TestMain(m *testing.M) {
+	// Start running the servers, but sleep for only 500ms
+	// because it should just take a couple ms for startup
+	// and install and run will take longer
+	go run(args{ConfigFile: "./settings.json"})
+	time.Sleep(500 * time.Millisecond)
+
+	err := playwright.Install()
+	if err != nil {
+		fmt.Printf("could not install playwright: %v\n", err)
+		os.Exit(1)
+	}
+
+	pw, err = playwright.Run()
+	if err != nil {
+		fmt.Printf("could not run playwright: %v\n", err)
+		os.Exit(1)
+	}
+
+	browsers = append(browsers, pw.Firefox)
+	browsers = append(browsers, pw.Chromium)
+
+	code := m.Run()
+	os.Exit(code)
+}
+
+func TestAccess(t *testing.T) {
+	runTests(t, func(t *testing.T, browser playwright.Browser) {
+		_, err := openChat(t, browser, "testUser")
+		if err != nil {
+			t.Error(err)
+		}
+	})
+}
+
+func TestChatMessage(t *testing.T) {
+	const msg = "testing 1 2 3"
+	runTests(t, func(t *testing.T, browser playwright.Browser) {
+		page1, err := openChat(t, browser, "testUser1")
+		if err != nil {
+			t.Errorf("could not open chat for user 1: %v", err)
```

Comment:
shouldn't these be `t.Fatalf()`?  If we get an error back here we can't exactly continue testing, right?

Reply:
Fixed in 28e698e59b442ed3997adc55d39fb54bb73d749d